lib/upipe-x265/upipe_x265.c

```diff
@@ -0,0 +1,1292 @@
+    return UBASE_ERR_NONE;
+}
+
+/** @internal @This receives a provided ubuf manager.
+ *
+ * @param upipe description structure of the pipe
+ * @param flow_format amended flow format
+ * @return an error code
+ */
+static int upipe_x265_check_ubuf_mgr(struct upipe *upipe,
+                                     struct uref *flow_format)
+{
+    struct upipe_x265 *upipe_x265 = upipe_x265_from_upipe(upipe);
+    if (flow_format == NULL)
+        return UBASE_ERR_NONE; /* should not happen */
+
+    uref_free(upipe_x265->flow_def_requested);
+    upipe_x265->flow_def_requested = flow_format;
+
+    bool was_buffered = !upipe_x265_check_input(upipe);
+    upipe_x265_output_input(upipe);
+    upipe_x265_unblock_input(upipe);
+    if (was_buffered && upipe_x265_check_input(upipe)) {
+        /* All packets have been output, release again the pipe that has been
+         * used in @ref upipe_x265_input. */
+        upipe_release(upipe);
+    }
+
+    return UBASE_ERR_NONE;
+}
+
+/** @internal @This sets the input flow definition.
+ *
+ * @param upipe description structure of the pipe
+ * @param flow_def flow definition packet
+ * @return false if the flow definition is not handled
+ */
+static int upipe_x265_set_flow_def(struct upipe *upipe,
+                                   struct uref *flow_def)
+{
+    if (flow_def == NULL)
+        return UBASE_ERR_INVALID;
+
+    /* We only accept YUV420P for the moment. */
+    uint8_t macropixel;
+    if (unlikely(!ubase_check(uref_flow_match_def(flow_def, EXPECTED_FLOW)) ||
+                 !ubase_check(uref_pic_flow_get_macropixel(flow_def, &macropixel)) ||
+                 macropixel != 1 ||
+                 !ubase_check(uref_pic_flow_check_chroma(flow_def, 1, 1, 1, "y8")) ||
+                 !ubase_check(uref_pic_flow_check_chroma(flow_def, 2, 2, 1, "u8")) ||
+                 !ubase_check(uref_pic_flow_check_chroma(flow_def, 2, 2, 1, "v8"))))
+        return UBASE_ERR_INVALID;
+
+    /* Extract relevant attributes to flow def check. */
+    struct uref *flow_def_check =
+        upipe_x265_alloc_flow_def_check(upipe, flow_def);
+    if (unlikely(flow_def_check == NULL)) {
+        upipe_throw_fatal(upipe, UBASE_ERR_ALLOC);
+        return UBASE_ERR_ALLOC;
+    }
+
+    struct urational fps;
+    uint64_t hsize, vsize;
+    if (!ubase_check(uref_pic_flow_get_fps(flow_def, &fps) ||
+        !ubase_check(uref_pic_flow_get_hsize(flow_def, &hsize)) ||
+        !ubase_check(uref_pic_flow_get_vsize(flow_def, &vsize)))) {
+        upipe_err(upipe, "incompatible flow def");
+        uref_free(flow_def_check);
+        return UBASE_ERR_INVALID;
+    }
+
+    if (unlikely(!ubase_check(uref_pic_flow_copy_format(flow_def_check, flow_def)) ||
+                 !ubase_check(uref_pic_flow_set_fps(flow_def_check, fps)) ||
+                 !ubase_check(uref_pic_flow_set_hsize(flow_def_check, hsize)) ||
+                 !ubase_check(uref_pic_flow_set_vsize(flow_def_check, vsize)))) {
```

Comment:
You need to take into account sar and overflow here, if you can't dynamically adjust.